test/integration/replicaset/replicaset_test.go

```diff
@@ -268,250 +510,438 @@ func TestAdoption(t *testing.T) {
-	rs.Spec.Template.Labels["uniqueKey"] = "1"
-	pod1 := newMatchingPod("pod1", ns.Name)
-	pod1.Labels["uniqueKey"] = "1"
-	pod2 := newMatchingPod("pod2", ns.Name)
-	pod2.Labels["uniqueKey"] = "2"
-	createRSsPods(t, clientSet, []*v1beta1.ReplicaSet{rs}, []*v1.Pod{pod1, pod2}, ns.Name)
+	rss, _ := createRSsPods(t, c, []*v1beta1.ReplicaSet{rs}, []*v1.Pod{})
+	rs = rss[0]
+	waitRSStable(t, c, rs)
+
+	podClient := c.Core().Pods(ns.Name)
+	pods := getPods(t, podClient, labelMap())
+	if len(pods.Items) != 1 {
+		t.Fatalf("len(pods) = %d, want 1", len(pods.Items))
+	}
+	pod := &pods.Items[0]
+	var falseVar = false
+
+	// When the only OwnerReference of the pod points to another type of API object such as statefulset
+	// with Controller=false, the RS should add a second OwnerReference (ControllerRef) pointing to itself
+	// with Controller=true
+	ownerReference := metav1.OwnerReference{UID: uuid.NewUUID(), APIVersion: "apps/v1beta1", Kind: "StatefulSet", Name: rs.Name, Controller: &falseVar}
+	testPodControllerRefPatch(t, c, pod, &ownerReference, rs, 2)
+
+	// When the only OwnerReference of the pod points to the RS, but Controller=false
+	ownerReference = metav1.OwnerReference{UID: rs.UID, APIVersion: "extensions/v1beta1", Kind: "ReplicaSet", Name: rs.Name, Controller: &falseVar}
+	testPodControllerRefPatch(t, c, pod, &ownerReference, rs, 1)
+}
 
-	stopCh := make(chan struct{})
-	informers.Start(stopCh)
-	go rm.Run(5, stopCh)
-	waitRSStable(t, clientSet, rs, ns.Name)
+func TestReadyAndAvailableReplicas(t *testing.T) {
+	s, closeFn, rm, informers, c := rmSetup(t)
+	defer closeFn()
+	ns := framework.CreateTestingNamespace("test-ready-and-available-replicas", s, t)
+	defer framework.DeleteTestingNamespace(ns, s, t)
+	defer close(runControllerAndInformers(rm, informers))
 
-	// change the rs's selector to match both pods
-	patch := `{"metadata":{"labels":{"uniqueKey":"1"}}}`
-	podClient := clientSet.Core().Pods(ns.Name)
-	pod2, err := podClient.Patch(pod2.Name, types.StrategicMergePatchType, []byte(patch))
+	rs := newRS("rs", ns.Name, 2)
+	rs.Spec.MinReadySeconds = 5
+	rss, _ := createRSsPods(t, c, []*v1beta1.ReplicaSet{rs}, []*v1.Pod{})
+	rs = rss[0]
+	waitRSStable(t, c, rs)
+
+	// First verify no pod is available
+	if rs.Status.AvailableReplicas != 0 {
+		t.Fatalf("Unexpected .Status.AvailableReplicas: Expected 0, saw %d", rs.Status.AvailableReplicas)
+	}
+
+	podClient := c.Core().Pods(ns.Name)
+	pods := getPods(t, podClient, labelMap())
+	if len(pods.Items) != 2 {
+		t.Fatalf("len(pods) = %d, want 2", len(pods.Items))
+	}
+	podNum := len(pods.Items)
+	// Separate 2 pods into their own list
+	firstPodList := &v1.PodList{Items: pods.Items[:podNum-1]}
+	secondPodList := &v1.PodList{Items: pods.Items[1:]}
+	// Mark all pods running and ready
+	// First pod's LastTransitionTime is 5 seconds ago
+	markPodsRunningAndReady(t, c, firstPodList, ns.Name)
+	// This is made possible by sleeping for 5 seconds
+	time.Sleep(time.Second * 5)
```

Comment:
It's best to avoid Sleep whenever possible. In this case, I think you can fake the LastTransitionTime by directly specifying it in the PodCondition that you build inside `markPodsRunningAndReady`. You may need to add a parameter to plumb that value in.

Reply:
I did try specifying it in the `PodCondition` but failed. IIRC, `UpdateStatus` changed `LastTransitionTime` to match the time the status is updated. I will find a way to plumb the value in using a parameter.

Reply:
I think this is only happening client-side in the call to `UpdatePodCondition()`:

https://github.com/kubernetes/kubernetes/blob/884c6f9146971f3d76a813b2f959a397591deddf/pkg/api/pod/util.go#L202-L203

You should be able to bypass this by avoiding UpdatePodCondition and instead directly implement just the parts of it that you want - i.e. call `GetPodReadyCondition` and either update (if it already exists) or append to `status.Conditions`.

Reply:
Thanks! I verify `GetPodReadyCondition` + update/append work!